pkg/controller/job/controller.go

```diff
@@ -255,6 +258,16 @@ func (jm *JobController) enqueueController(obj interface{}) {
 	// all controllers there will still be some replica instability. One way to handle this is
 	// by querying the store for all controllers that this rc overlaps, as well as all
 	// controllers that overlap this rc, and sorting them.
+	// Instead of insert the key directly, check if this key is in the queue already
+	// If it isn't or this key is in the head of the queue, insert the key, otherwise, abort.
+	// If the controller fail and restart, both queue and queuedJob will be lost
+	// everything need to be start over.
+	if jm.queuedJob.Has(key) {
```

Comment:
Thanks @smarterclayton , you are right. The workqueue has already done this. It is my bad.

But I think the Add() func in workqueue should not check if the item is already in q.processing. If it in, we still need to insert this item into our queue. Since we insert the item into the q.processing and remove it from q.queue when we call Get(). This means items in q.processing is already out of queue.

This code is a good example why we should still insert: the processing item can be in any stage of synchronization. If this synchronization comes to end but have not being removed from q.processing(haven't call Done()), the new update which calls q.Add() should still add the controller into the queue.